library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -1,65 +1,216 @@
+  -- | The arguments to the application.
   [ErasedAnn.AnnTerm PrimTy RawPrimVal] ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-mkApp f (ErasedAnn.PrimTy ty) xs = case f of
-  ErasedAnn.Prim prim -> applyPrim prim ty xs
+  m LLVM.Operand
+mkApp env f@(ErasedAnn.Ann {ErasedAnn.term, ErasedAnn.type'}) _ xs =
+  case term of
+    ErasedAnn.LamM {ErasedAnn.body, ErasedAnn.arguments, ErasedAnn.capture} -> do
+      funname <- mkLam env type' body arguments capture
+      xs' <- mapM (compileTerm env) xs
+      let xs'args = zip xs' (repeat [])
+      LLVM.call (globalRef (typeToLLVM type') funname) xs'args
+    ErasedAnn.Prim prim -> applyPrim env prim xs
+    ErasedAnn.Var v -> do
+      f' <- compileTerm env f
+      xs' <- mapM (compileTerm env) xs
+      let xs'args = zip xs' (repeat []) -- Do not pass attributes to the args.
+      LLVM.call f' xs'args
+
+-- | Write LLVM code for a primitive.
+-- TODO: implement other primitives.
+mkPrim ::
+  LLVM.MonadIRBuilder m =>
+  -- | Term that contains the primitive.
+  ErasedAnn.Term PrimTy RawPrimVal ->
+  -- | Type of the primitive.
+  ErasedAnn.Type PrimTy ->
+  m LLVM.Operand
+mkPrim (ErasedAnn.Prim prim) ty = case prim of
+  LitInt i -> case ty of
+    ErasedAnn.PrimTy (PrimTy LLVM.IntegerType {LLVM.typeBits}) ->
+      return $
+        LLVM.ConstantOperand $
+          LLVM.Int {LLVM.integerBits = typeBits, LLVM.integerValue = i}
 
+-- | Generate code for primitives that are used as a function.
+-- TODO: implement other primitives.
 applyPrim ::
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of global variables.
+  Env ->
+  -- | The function primitive of the application.
   RawPrimVal ->
-  PrimTy ->
+  -- | The arguments to the application.
   [ErasedAnn.AnnTerm PrimTy RawPrimVal] ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-applyPrim f ty xs
+  m LLVM.Operand
+applyPrim env f xs
   | arityRaw f == lengthN xs =
     case f of
       Add -> do
-        x <- compileTerm (xs P.!! 0)
-        y <- compileTerm (xs P.!! 1)
+        x <- compileTerm env (xs P.!! 0)
+        y <- compileTerm env (xs P.!! 1)
         LLVM.add x y
 
+-- | Counterpart of `mkName`: make a `ParameterName` given a name.
+mkParameterName :: NameSymbol.T -> LLVM.ParameterName
+mkParameterName s = S.fromString $ unintern $ NameSymbol.toSymbol s
+
+-- | Handy wrapper for creating global references based on a type and name.
+globalRef :: LLVM.Type -> LLVM.Name -> LLVM.Operand
+globalRef ty name = LLVM.ConstantOperand $ LLVM.GlobalReference ty name
+
+-- | Translate a Juvix type into an LLVM type.
 typeToLLVM :: ErasedAnn.Type PrimTy -> LLVM.Type
 typeToLLVM (ErasedAnn.PrimTy (PrimTy ty)) = ty
+typeToLLVM (ErasedAnn.Pi _usage f xs) =
+  LLVM.FunctionType
+    { LLVM.resultType = typeToLLVM resultType,
+      LLVM.argumentTypes = map typeToLLVM argumentTypes,
+      LLVM.isVarArg = False
+    }
+  where
+    tyList = f : functionTy xs
+    resultType = P.last tyList
```

Comment:
can this fail?

Reply:
No, `tyList` has atleast two elements, and therefore we can always execute `last` and `init` on it.